pycalphad/core/conditions.py

```diff
@@ -0,0 +1,182 @@
+            return [float(tup[0])]
+        elif len(tup) == 2:
+            return np.arange(tup[0], tup[1], dtype=np.float64)
+        elif len(tup) == 3:
+            return np.arange(tup[0], tup[1], tup[2], dtype=np.float64)
+        else:
+            raise ValueError('Condition tuple is length {}'.format(len(tup)))
+    elif isinstance(tup, Q_):
+        return tup
+    elif isinstance(tup, Iterable) and np.ndim(tup) != 0:
+        return [float(x) for x in tup]
+    else:
+        return [float(tup)]
+
+class Conditions:
+    _wks: "Workspace"
+    _conds: List[ConditionsEntry]
+
+    minimum_composition: float = 1e-10
+
+    def __init__(self, wks: Optional["Workspace"]):
+        self._wks = wks
+        self._conds = []
+        # Default to N=1
+        self.__setitem__(v.N, Q_(np.atleast_1d(1.0), 'mol'))
+
+    @classmethod
+    def from_dict(cls, d):
+        if isinstance(d, Conditions):
+            return d
+        obj = cls(wks=None)
+        obj.update(d)
+        return obj
+    
+    def _find_matching_index(self, prop: "ComputableProperty"):
+        for idx, (key, _) in enumerate(self._conds):
+            # TODO: Use more sophisticated matching
+            if str(prop) == str(key):
+                return idx
+        return None
+
+    @classmethod
+    def cast_from(cls, key) -> "Conditions":
+        return cls.from_dict(key)
+    
+    def __getitem__(self, item):
+        key = as_property(item)
+        idx = self._find_matching_index(key)
+        if idx is None:
+            raise IndexError(f"{item} is not a condition")
+        entry = self._conds[idx]
+        # Important to use the _key_ display_units, and not the entry.prop
+        # This is because v.T['K'] == v.T['degC'], so conditions can be
+        # stored and queried with distinct units
+        return entry.value.to(key.display_units).magnitude
+
+    def get(self, item, default=_default):
+        try:
+            return self.__getitem__(item)
+        except IndexError:
+            if default is not _default:
+                return default
+            else:
+                raise
+
+    def __delitem__(self, item):
+        idx = self._find_matching_index(as_property(item))
+        if idx is None:
+            raise IndexError(f"{item} is not a condition")
+        del self._conds[idx]
+    
+    def __setitem__(self, item, value):
+        prop = as_property(item)
+        if isinstance(prop, (v.MoleFraction, v.MassFraction, v.SiteFraction)):
+            vals = unpack_condition(value)
+            if isinstance(vals, Q_):
+                vals = vals.to(prop.implementation_units).magnitude
+            # "Zero" composition is a common pattern. Do not warn for that case.
+            if np.any(np.logical_and(np.asarray(vals) < self.minimum_composition, np.asarray(vals) > 0)):
+                warnings.warn(
+                    f"Some specified compositions are below the minimum allowed composition of {self.minimum_composition}.")
```

Comment:
1.0-epsilon is the de facto maximum allowed composition today, but I don't think it needs its own warning here because `v.X(A) = 1 - [tiny value]` is a much less common pattern than `v.X(A) = [tiny value]`.